src/Npgsql/MultiHostConnectorPool.cs

```diff
@@ -72,17 +74,19 @@ static ClusterState GetClusterState(string host, int port, bool ignoreExpiration
             => ClusterStateCache.GetClusterState(host, port, ignoreExpiration);
 
         async ValueTask<NpgsqlConnector?> TryGetIdleOrNew(NpgsqlConnection conn, TimeSpan timeoutPerHost, bool async,
-            TargetSessionAttributes preferredType, Func<ClusterState, TargetSessionAttributes, bool> clusterValidator, int poolIndex,
+            TargetSessionAttributes preferredType, Func<ClusterState, TargetSessionAttributes, bool> clusterValidator,
             IList<Exception> exceptions, CancellationToken cancellationToken)
         {
             var pools = _pools;
+            var poolIndex = -1;
             for (var i = 0; i < pools.Length; i++)
             {
-                var pool = pools[poolIndex];
                 poolIndex++;
                 if (poolIndex == pools.Length)
                     poolIndex = 0;
 
+                poolIndex = conn.Settings.LoadBalanceHosts ? GetRoundRobinIndex() : poolIndex;
```

Comment:
Doesn't it mean, that with an enabled LoadBalance in some cases I might get an exception, because `GetRoundRobinIndex` returns one and the same host?

Reply:
> Am I wrong? Is there an oversight?

Concurrent `Open` from multiple threads.

Reply:
> Concurrent `Open` from multiple threads.

You are right but isn't that a general problem that can also happen with `GetRoundRobinIndex()` at the previous location (in `Get()`)?

Reply:
`Get` is only called once per `Open`, so we also get the starting index only once.

> (we could make _roundRobinIndex an ulong btw)

Definitely should, that's an oversight on my part)

Reply:
> Definitely should, that's an oversight on my part)

That actually simplifies the code a lot.

No real solution for the other problem yet, though.

Reply:
> No real solution for the other problem yet, though.

I do have one idea: whenever we get an index, we check whether the previous host (index - 1) is still alive. If it is, nothing really changes. If it's not, we have to get another round-robin index and make sure it differs from the one we have. It doesn't solve everything (at the very least there is `Prefer` can of worms to deal with), but that's a start.

Reply:
Ok, I see that this is ways more complicated than I initially thought.
Maybe even beyond my capabilities.
I'll remove this commit for now and update the issue.
You can give it a shot if you want to.